Mathlib/Topology/ProperMap.lean

```diff
@@ -136,6 +126,43 @@ lemma IsProperMap.ultrafilter_le_nhds_of_tendsto (h : IsProperMap f) ⦃𝒰 : U
     (hy : Tendsto f 𝒰 (𝓝 y)) : ∃ x, f x = y ∧ 𝒰 ≤ 𝓝 x :=
   (isProperMap_iff_ultrafilter.mp h).2 hy
 
+/-- The composition of two proper maps is proper. -/
+lemma IsProperMap.comp (hf : IsProperMap f) (hg : IsProperMap g) :
+    IsProperMap (g ∘ f) := by
+  refine ⟨by continuity, fun ℱ z h ↦ ?_⟩
+  rw [MapClusterPt, ← map_map] at h
+  rcases hg.clusterPt_of_mapClusterPt h with ⟨y, rfl, hy⟩
+  rcases hf.clusterPt_of_mapClusterPt hy with ⟨x, rfl, hx⟩
+  use x, rfl
+
+/-- If the composition of two continuous functions `g ∘ f` is proper and `f` is surjective,
+then `g` is proper. -/
+lemma isProperMap_of_comp_of_surj (hf : Continuous f)
+    (hg : Continuous g) (hgf : IsProperMap (g ∘ f)) (f_surj : f.Surjective) : IsProperMap g := by
+  refine ⟨hg, fun ℱ z h ↦ ?_⟩
+  rw [← ℱ.map_comap_of_surjective f_surj, MapClusterPt, map_map, ← MapClusterPt] at h
+  rcases hgf.clusterPt_of_mapClusterPt h with ⟨x, rfl, hx⟩
+  rw [← ℱ.map_comap_of_surjective f_surj]
+  exact ⟨f x, rfl, hx.map hf.continuousAt tendsto_map⟩
```

Comment:
In this lemma and the previous one, there is too much definitional rewriting going on. This is the signal that a stupid lemma is missing. Something like:
```lean
lemma mapClusterPt_comp {X α β : Type*} {x : X} [TopologicalSpace X] {F : Filter α} {φ : α → β}
    {u : β → X} : MapClusterPt x F (u ∘ φ) ↔ MapClusterPt x (map φ F) u := Iff.rfl
```
that should go in an earlier file, very close to the definition of `MapClusterPt`.